planner/solver.go

```diff
@@ -282,7 +309,84 @@ func reversePlans(plans []PlanningSolution) []PlanningSolution {
 	return plans
 }
 
-func (s *Solver) generateSolutions(ctx context.Context, req *PlanningReq, timeMatcher matching.Matcher, priceRangeMatcher matching.Matcher) (resp PlanningResp) {
+func (s *Solver) FindOptimalPlan(placeClusters [][]matching.Place) ([]string, error) {
+	placeIds, weights, err := s.weightMatrix(placeClusters)
+	if err != nil {
+		return nil, err
+	}
+	solve, err := hungarianAlgorithm.Solve(fixWeights(weights))
+	if err != nil {
+		return nil, err
+	}
+	var result []string
+	for _, idx := range solve[:len(placeClusters)] {
+		result = append(result, placeIds[idx])
+	}
+	return result, nil
+}
+
+// We are solving a maximization problem instead of the default minimization problem Hungarian algorithm is designed for.
+// Therefore, needs to use the maximum value to minus all the values in the matrix.
+func fixWeights(weights [][]int) [][]int {
+	if len(weights) == 0 {
+		return nil
+	}
+	fixedWeights := make([][]int, len(weights))
+	for idx := range fixedWeights {
+		fixedWeights[idx] = make([]int, len(weights[0]))
+	}
+	// assume the values are non-negative
+	var maxValue int
+	for _, ws := range weights {
```

Comment:
nit: if Hungarian algorithm supports negative value in the weight matrix, we can skip finding the maxValue here and use `-val` directly?

Reply:
It appears that using negated values do not equivalent to solving a maximization problem.